#142 - Support Redis Cluster

yergo opened the pull request:
*  Type: new feature
*  Link to issue: https://github.com/phalcon/phalcon/issues/142

**In raising this pull request, I confirm the following:**

-  [x] I have read and understood the [Contributing Guidelines](https://github.com/phalcon/phalcon/blob/master/CONTRIBUTING.md)
-  [x] I have checked that another pull request for this purpose does not exist
-  [x] I wrote some tests for this PR
-  [ ] I have updated the relevant CHANGELOG
-  [ ] I have created a PR for the [documentation](https://github.com/phalcon/docs) about this change

Small description of change:
Added Storage\Adapter\RedisCluster for \RedisCluster to be compatible with framework.

yergo committed:
#142 - Support Redis Cluster

Added RedisCluster.php adapter

niden commented on src/Storage/Adapter/RedisCluster.php:
```diff
@@ -0,0 +1,315 @@
+    /**
+     * @var string
+     */
+    protected string $prefix = 'ph-reds-';
```
This is the same key as the Redis adapter. Can we change this please? Perhaps `ph-redc-` (for cluster)

yergo replied:
changed

niden commented on src/Storage/Adapter/RedisCluster.php:
```diff
@@ -0,0 +1,315 @@
+ *
+ * @property array $options
+ */
+class RedisCluster extends AbstractAdapter
```
It might be best to extend the `Redis` adapter. This way certain methods (such as `decrement` for example) are the same - > less code :)

yergo replied:
Actually, `select`, `checkConnect`, `checkAuth` and `checkIndex` - that were removed, were private.... yeah, I'll update

niden replied:
If you need to, make them protected in `Redis` to reuse them. Just a thought.

yergo replied:
it tricked me a little, I was producing $options with configured in defaults from Adapter\Redis, but after small refactoring it is ok now.

niden commented on src/Storage/Adapter/RedisCluster.php:
```diff
@@ -0,0 +1,315 @@
+<?php
```
We will need some tests for this. We need to set up a redis cluster in the docker environment as a service to be able to run tests on it. I will help with that so reach out on discord.

yergo replied:
I'm on solo challenge, but appreciate. I have seen already codeception integration tests and how docker tests are run, hopefully can do this on my own and learn something new doing this.

yergo replied:
I added Cache\RedisCluster adapter and implemented almost all integration tests there. Still 4 tests that require maybe a little refactoring to kepp them DRY. Aiming for adjusting test in Storage\Adapter namespace too

yergo replied:
sonarcloud claim 12% duplication code in tests i only adjusted a bit. Don't know what should I do with this, code marked as duplicate is not the code I wrote

niden replied:
Don't worry about the duplication of code in the tests.

yergo committed:
#142 - Support Redis Cluster

RedisCluster Adapter to inherit from Redis Adapter

yergo committed:
#142 - Support Redis Cluster

Codacy post-static analysis adjustments

yergo committed:
#142 - Support Redis Cluster

RedisCluster as Cache adapter

yergo committed:
#142 - Support Redis Cluster

- docker cluster configuration
- common Cache/Adapter/Redis RedisCluster tests
- improvements in class inheritance configuration separation
- old issue with consts not set in RedisCluster phpredis library worked around

yergo committed:
#142 - Support Redis Cluster

- revert unnecesary formatting update from Redis.php
- added missing newline for RedisCluster.php

yergo committed:
#142 - Support Redis Cluster

adjust to codestyle

yergo committed:
#142 - Support Redis Cluster

remove duplicated methods

yergo committed:
#142 - Support Redis Cluster

remove obsolete imports

yergo committed:
#142 - Support Redis Cluster

update Cache\Redis integration tests

yergo committed:
#142 - Support Redis Cluster

update integration tests for Storage\Adapter to include testing RedisCluster adapter

niden commented on src/Storage/Adapter/RedisCluster.php:
```diff
@@ -0,0 +1,208 @@
+     * } $options
+     * @throws SupportException
+     */
+    public function __construct(SerializerFactory $factory, array $options = [])
```
This is not really necessary. The parent will be called either way.

niden commented:
Than kyou @yergo
